### PR TITLE
CEDS-3025 Reply to message

### PR DESCRIPTION
**Pull request by udinobi**
Page showing the selected conversation including a reply box and page showing the reply result

### CODE REVIEW

File: app/controllers/SecureMessagingController.scala [hunk 1]
```diff
@@ -34,24 +38,50 @@ class SecureMessagingController @Inject()(
   messageConnector: SecureMessageFrontendConnector,
   mcc: MessagesControllerComponents,
   inbox_wrapper: inbox_wrapper,
-  conversation_wrapper: conversation_wrapper
+  conversation_wrapper: conversation_wrapper,
+  reply_result: reply_result
```
Comment (glidester) [inline app/controllers/SecureMessagingController.scala:42]: Change name to 'reply_wrapper' to keep consistency with other two view names

File: app/forms/ReplyToMessage.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+package forms
+
+import play.api.data.Form
+import play.api.data.Forms.{mapping, nonEmptyText}
+
+case class ReplyToMessage(messageReply: String)
+
+object ReplyToMessage {
+
+  val form: Form[ReplyToMessage] = Form(mapping("messageReply" -> nonEmptyText)(ReplyToMessage.apply)(ReplyToMessage.unapply))
```
Comment (glidester) [inline app/forms/ReplyToMessage.scala:26]: Not sure we should be doing any validation like this 'nonEmptyText' validation. DC will do this at their end and return to us a partial with the form error per-rendered.

File: app/controllers/SecureMessagingController.scala (hunk 1 above)
Comment (udinobi) [reply app/controllers/SecureMessagingController.scala:42]: :+1:

File: app/forms/ReplyToMessage.scala (hunk 2 above)
Comment (udinobi) [reply app/forms/ReplyToMessage.scala:26]: This is just temporary. Let's wait to actually know what DC expect in the payload of the 'POST Reply'.

Comment (glidester) [reply app/forms/ReplyToMessage.scala:26]: Just got off a call with DC and they confirmed that they will do the validation their end and return the conversation partial with any form errors rendered as part of the POST's response. So we definitely will not need this validation our end.

Comment (udinobi) [reply app/forms/ReplyToMessage.scala]: Ok, removed.

**Review by glidester — APPROVED** [review]

File: app/controllers/SecureMessagingController.scala [hunk 3]
```diff
@@ -34,24 +36,46 @@ class SecureMessagingController @Inject()(
...
+
+  def submitReply(client: String, conversationId: String): Action[AnyContent] = actions.async { implicit request =>
+    request.body.asFormUrlEncoded.map { reply =>
+      messageConnector.submitReply(client, conversationId, reply)
     }
+
+    // For the time being... until the downstream service is ready
+    Future(Redirect(routes.SecureMessagingController.displayReplyResult(client, conversationId)))
+
+  // Future.successful(NoContent)
```
Comment (MaciejRewera) [inline app/controllers/SecureMessagingController.scala:74]: This seems not needed (?)

File: app/controllers/SecureMessagingController.scala [hunk 4]
```diff
@@ -34,24 +36,46 @@ class SecureMessagingController @Inject()(
...
+  }
+
+  private def wrapperFormForPartial(partial: ConversationPartial)(implicit request: Request[_]): HtmlFormat.Appendable = {
+    val csrfToken = CSRF.getToken.get.value
+    conversation_wrapper(HtmlFormat.raw(partial.body.replace("[CSRF_TOKEN_TO_REPLACE]", csrfToken)))
+  }
```
Comment (MaciejRewera) [inline app/controllers/SecureMessagingController.scala:80]: This method is only useful for one of the pages. Was it the intention, or should it be more generic to handle other pages as well?

File: app/controllers/SecureMessagingController.scala [hunk 5]
```diff
@@ -34,24 +36,46 @@ class SecureMessagingController @Inject()(
...
     }
+
+    // For the time being... until the downstream service is ready
+    Future(Redirect(routes.SecureMessagingController.displayReplyResult(client, conversationId)))
+
+  // Future.successful(NoContent)
+  }
+
+  private def wrapperFormForPartial(partial: ConversationPartial)(implicit request: Request[_]): HtmlFormat.Appendable = {
+    val csrfToken = CSRF.getToken.get.value
```
Comment (MaciejRewera) [inline app/controllers/SecureMessagingController.scala:78]: Please don't use `get` on `Option` in production code.

File: test/controllers/SecureMessagingControllerSpec.scala [hunk 6]
```diff
@@ -137,14 +144,97 @@ class SecureMessagingControllerSpec extends ControllerSpecBase with TestRequests
...
+
+            val result = controller.displayReplyResult(clientId, conversationId)(fakeRequest)
+
+            status(result) mustBe OK
+          }
+        }
+
+        "unsuccessfully returns a failed Future" should {
+
+          "display the 'Sorry' page to the user" in {
```
Comment (MaciejRewera) [inline test/controllers/SecureMessagingControllerSpec.scala:193]: Nit: This text is not aligned with what you assert against. It says about displaying 'Sorry' page, but assertion is against an Exception being thrown.

File: app/controllers/SecureMessagingController.scala (hunk 5 above)
Comment (udinobi) [reply app/controllers/SecureMessagingController.scala:78]: This is not production code, it is just temporary. We won't need to do this when DC is ready. They will also provide the CSRF Token that we do not have currently in the prototype.

File: app/controllers/SecureMessagingController.scala (hunk 4 above)
Comment (udinobi) [reply app/controllers/SecureMessagingController.scala:80]: This is just temporary. We won't need to do this when DC is ready. They will also provide the CSRF Token that we do not have currently in the prototype.

File: app/controllers/SecureMessagingController.scala (hunk 3 above)
Comment (udinobi) [reply app/controllers/SecureMessagingController.scala:74]: This is more a sort of question/note. I'm not sure what response we will have to send here... the connector's response? NoContent? or?

**Pull request merged**